pymongo_auth_aws/auth.py

```diff
@@ -93,124 +92,40 @@ def dst(self, dt):
-                'error: %s' % (exc,))
-    else:
-        # If the environment variable AWS_CONTAINER_CREDENTIALS_RELATIVE_URI is
-        # not set drivers MUST assume we are on an EC2 instance and use the
-        # endpoint
-        # http://169.254.169.254/latest/meta-data/iam/security-credentials
-        # /<role-name>
-        # whereas role-name can be obtained from querying the URI
-        # http://169.254.169.254/latest/meta-data/iam/security-credentials/.
-        try:
-            # Get token
-            headers = {'X-aws-ec2-metadata-token-ttl-seconds': "30"}
-            res = requests.put(_AWS_EC2_URI+'latest/api/token',
-                               headers=headers, timeout=_AWS_HTTP_TIMEOUT)
-            token = res.content
-            # Get role name
-            headers = {'X-aws-ec2-metadata-token': token}
-            res = requests.get(_AWS_EC2_URI+_AWS_EC2_PATH, headers=headers,
-                               timeout=_AWS_HTTP_TIMEOUT)
-            role = res.text
-            # Get temp creds
-            res = requests.get(_AWS_EC2_URI+_AWS_EC2_PATH+role,
-                               headers=headers, timeout=_AWS_HTTP_TIMEOUT)
-            res_json = res.json()
-        except (ValueError, requests.exceptions.RequestException) as exc:
-            raise PyMongoAuthAwsError(
-                'temporary MONGODB-AWS credentials could not be obtained, '
-                'error: %s' % (exc,))
+    creds = get_cached_credentials()
+    if creds:
+        return creds
 
-    # See https://docs.aws.amazon.com/cli/latest/reference/sts/assume-role.html#examples for expected result format.
     try:
-        temp_user = res_json['AccessKeyId']
-        temp_password = res_json['SecretAccessKey']
-        session_token = res_json['Token']
-        expiration = res_json['Expiration']
-    except KeyError:
+        session = boto3.Session()
+        creds = session.get_credentials()
+        # Use frozen credentials to prevent a race condition if there
+        # is a refresh between property accesses.
+        frozen = creds.get_frozen_credentials()
+    except Exception:
         # If temporary credentials cannot be obtained then drivers MUST
         # fail authentication and raise an error.
+        set_cached_credentials(None)
         raise PyMongoAuthAwsError(
             'temporary MONGODB-AWS credentials could not be obtained')
 
-    creds = AwsCredential(temp_user, temp_password, session_token, expiration)
-    if use_cached_credentials:
+    # The botocore Credentials object does not expose the expiration
+    # directly, instead we use the refresh_needed method to determine
+    # whether the credentials are expired.
+    refresh_needed = getattr(creds, 'refresh_needed', None)
+    creds = AwsCredential(
+        frozen.access_key, frozen.secret_key, frozen.token, refresh_needed
+    )
+    # Only cache credentials that need to be refreshed from
+    # an external source.
+    if refresh_needed is not None:
```

Comment:
Can we add a test to make sure we don't cache env var lookup creds?

Reply:
Done